### PR TITLE
feat: scan argo cd instances for app

### LINKED ISSUE #193: ArgoCD instance per cluster
We have one argocd instance per cluster. The proxy solution mentioned doesn't seem scalable to us. We'd also need to add a path on catalog files for this to work properly.

The current kubernetes plugin finds an app across all clusters based on selectors. Could this plugin do something similar instead of having a proxy path added to each and every catalog-info file?

I started writing code to take over the argocd proxy URL on the backend to redirect the traffic based on what cluster an app resides on, but argocd plugin doesn't send cluster info making this harder and we'd rather work with you to make a solution than working in a silo.

All our argocd instances have URLs based on a pattern. Given the cluster name, we can determine the argocd instance URL.

### PR DESCRIPTION
**Pull request by karlhaworth**
From original PR of https://github.com/RoadieHQ/roadie-backstage-plugins/pull/208

This brings in a frontend enhancement to scan for argo apps rather than defining what proxy instance it's attached to. This is similar to how the backstage kubernetes plugin works.

This brings in a new backend plugin as well to support scanning instances on the backend for an app, verses using a proxy on the frontend.

#### :heavy_check_mark: Checklist

- [X] Added tests for new functionality and regression tests for bug fixes
- [ ] Screenshots of before and after attached (for UI changes)
- [X] Added or updated documentation (if applicable)

Resolves #193

### CODE REVIEW

**Comment by karlhaworth** [comment]
I still need to work out one test that was failing of `should display empty table when no item returned with app selector`

**Comment by karlhaworth** [comment]
@alzafacon assisted with fixing of tests. Thank you!

File: plugins/backend/backstage-plugin-argo-cd-backend/src/service/argocd.service.ts
Comment (Xantier) [inline plugins/backend/backstage-plugin-argo-cd-backend/src/service/argocd.service.ts:101]: Let's remove this argument and move the burden of adding/omitting the 'Bearer' prefix to the token to the user when they configure these on their app-config.

File: plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts:137]: Small thing, but could you rename this to `serviceLocatorUrl` or `ArgoBackendUrl` or something. It will always call the backend portion of this plugin in this case right?

File: plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDDetailsCard.tsx
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDDetailsCard.tsx:80]: This is failing type check it looks like. Unused

File: plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDHistoryCard.tsx
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDHistoryCard.tsx:48]: This unused as well.

File: plugins/frontend/backstage-plugin-argo-cd/config.d.ts
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/config.d.ts:22]: Are these actually ever needed on the frontend? We only care that the `appLocatorMethods` exists and is not empty right? We could possibly omit the `@visibility frontend` from these subitems or omit them completely, leaving them only present in the backend `config.d.ts`. I'm not sure if it changes the boolean check logic but I feel it might be fine.

File: plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts:124]:
```suggestion
    if (this.searchInstances) {
```

File: plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts:58]:
```suggestion
    if (this.searchInstances) {
```

File: plugins/backend/backstage-plugin-argo-cd-backend/package.json
Comment (Xantier) [inline plugins/backend/backstage-plugin-argo-cd-backend/package.json:6]: When we publish this NPM will handle linking etc. stuff based on the repository information. Could add a block for those in here. Something like:

  "repository": {
    "type": "git",
    "url": "github:RoadieHQ/roadie-backstage-plugins",
    "directory": "plugins/backend/backstage-plugin-argo-cd-backend"
  },

File: plugins/backend/backstage-plugin-argo-cd-backend/package.json
Comment (Xantier) [inline plugins/backend/backstage-plugin-argo-cd-backend/package.json:29]: Looks like `axios` is used as the XHR client for all request so you should be able to leave this out. Same for the global fetch definition in `plugins/backend/backstage-plugin-argo-cd-backend/src/setupTests.ts`

File: plugins/backend/backstage-plugin-argo-cd-backend/src/service/router.ts
Comment (Xantier) [inline plugins/backend/backstage-plugin-argo-cd-backend/src/service/router.ts:1]: Probably better to leave this notice out for now since other files don't have it either. We'll do a bulk update of these notices at a later stage.

File: plugins/frontend/backstage-plugin-argo-cd/README.md
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/README.md:133]: I'm not sure if this will link correctly in npmjs.com but we can modify it after shipping if that is the case.

File: plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts]: addressed

File: plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts]: addressed

File: plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDHistoryCard.tsx
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDHistoryCard.tsx]: addressed

File: plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDDetailsCard.tsx
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/src/components/ArgoCDDetailsCard.tsx]: addressed

File: plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/src/api/index.ts:137]: addressed

**Comment by Xantier** [comment]
Thank you so much @karlhaworth ! 

This looks amazing! There are few niggling things but other than that I think it looks to be ready to be shipped. I added few comments inline in the code and there are few additional items outside of the changeset that could be good to have:
* could you add a `config.d.ts` file into `plugins/backend/backstage-plugin-argo-cd-backend`
  * The contents can mimic `plugins/frontend/backstage-plugin-argo-cd/config.d.ts` unless there are some items that are not needed on the BE. 
  * Remember to add a line pointing to that file in `package.json` `dist` block as well
* Could you modify the README.md at the root to link to these plugins with attributions to AA

I think after those items this should probably be good to be shipped. After merging I will also modify the GitHub project to have a Codeowners file so you should be able to have more permissions into these packages directly.

File: plugins/backend/backstage-plugin-argo-cd-backend/src/service/argocd.service.ts
Comment (karlhaworth) [reply plugins/backend/backstage-plugin-argo-cd-backend/src/service/argocd.service.ts]: sounds good! I think this was leftover from our code.

File: plugins/backend/backstage-plugin-argo-cd-backend/package.json
Comment (karlhaworth) [reply plugins/backend/backstage-plugin-argo-cd-backend/package.json:6]: addressed

File: plugins/backend/backstage-plugin-argo-cd-backend/package.json
Comment (karlhaworth) [reply plugins/backend/backstage-plugin-argo-cd-backend/package.json]: addressed

File: plugins/backend/backstage-plugin-argo-cd-backend/src/service/router.ts
Comment (karlhaworth) [reply plugins/backend/backstage-plugin-argo-cd-backend/src/service/router.ts]: addressed.

**Comment by Xantier** [comment]
Could you also bump the version of the frontend plugin with a minor.

File: plugins/frontend/backstage-plugin-argo-cd/config.d.ts
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/config.d.ts]: addressed. seems to be working just fine. tests are passing and its working locally still.

**Comment by karlhaworth** [comment]
> Could you also bump the version of the frontend plugin with a minor.

`1.2.1` ? Doing so

**Comment by Xantier** [comment]
I think `1.2.0` should be the next minor.

File: plugins/frontend/backstage-plugin-argo-cd/README.md
Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/README.md:133]: 👍🏻 if I need to modify now happy to do so! Not sure what to change it to. I'm fine waiting too!

**Comment by karlhaworth** [comment]
> I think `1.2.0` should be the next minor.

right. not sure what I was thinking.

**Comment by karlhaworth** [comment]
> could you add a config.d.ts file into plugins/backend/backstage-plugin-argo-cd-backend

done with package.json updates

> Could you modify the README.md at the root to link to these plugins with attributions to AA

I am not sure exactly what you mean to this. The below?

```
 [Argo CD](https://www.npmjs.com/package/@roadiehq/backstage-plugin-argo-cd).

    - Feature: Scanning of all Argo CD instances for an app contributed by [American Airlines](https://github.com/AmericanAirlines)

 [Argo CD Backend](https://www.npmjs.com/package/@roadiehq/backstage-plugin-argo-cd-backend).

    - Contributed by [American Airlines](https://github.com/AmericanAirlines)
 ```

**Comment by Xantier** [comment]
> I am not sure exactly what you mean to this. The below?

Yup something like that looks good. Maybe the first one with something like:
'In collaboration with American Airlines' or similar.

**Comment by karlhaworth** [comment]
Ok for now I think I've made all the requested updates! 😃 

Let me know what else is needed! I appreciate your time.

**Comment by Xantier** [comment]
Thanks @karlhaworth!

Seems to be some failures on e2e tests still. I'll run few cycles on this locally and if all looks good, this should be good to be merged in after the CI passes.

File: app-config.yaml
Comment (Xantier) [inline app-config.yaml:154]: If you modify this to be an empty array, it should revert back to the old functionality of using the proxy for ArgoCD instead and the old e2e tests should succeed.
```suggestion
  appLocatorMethods: []
```

Maybe add the lines below in as a comment instead.

Comment (karlhaworth) [reply app-config.yaml]: trying that now.

Comment (karlhaworth) [reply app-config.yaml]: we don't use cypress internally at this moment however a PR is coming to enable it after our teams went through training. I am running that locally now as well for testing.Thanks for the suggestion.

Comment (Xantier) [reply app-config.yaml]: Yup, I think this change and the config boolean check on the frontend side should revert to the original proxy approach for this plugin --> E2E tests should pass like before.

File: plugins/frontend/backstage-plugin-argo-cd/src/components/useAppDetails.ts
Comment (Xantier) [inline plugins/frontend/backstage-plugin-argo-cd/src/components/useAppDetails.ts:38]: Here's the other location where we should check the `length` of the config array to be > 0

File: plugins/frontend/backstage-plugin-argo-cd/src/plugin.ts
Comment (Xantier) [reply plugins/frontend/backstage-plugin-argo-cd/src/plugin.ts]: Nice. I kicked off the CI process now and will take another look tomorrow morning at these to get this over the line.

Comment (Xantier) [reply plugins/frontend/backstage-plugin-argo-cd/src/plugin.ts]: We'll make a note to write better automation/documentation for the local runs of Cypress :+1:

Comment (karlhaworth) [reply plugins/frontend/backstage-plugin-argo-cd/src/plugin.ts]: 👍🏻 it could be me. It seems that the cmd is `yarn dev:ci` and it's running. Every once in a while I see lots of activity, but it doesn't seem to go anywhere? It's been running for quite a while.

**Review by Xantier — APPROVED** [review]
E2E looks to be in better shape now. :+1: 

We'll make a release out of this tomorrow and can continue to iterate on top of that then if needed.

Once again thanks for the contribution!

**Comment by karlhaworth** [comment]
Amazing! I look forward to this being released.

**Comment by Xantier** [comment]
:shipit:

**Comment by abkura** [comment]
@karlhaworth is secure: false applicable for multi-instance configuration?

**Pull request merged**
